ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -231,6 +261,42 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
     }
   }
 
+  /**
+   * Generate the partition override map containing partition state from all datacenters.
+   */
+  private Map<String, Map<String, String>> generatePartitionOverrideFromAllDCs() {
+    Map<String, Map<String, String>> partitionOverrideInfos = new HashMap<>();
+    for (PartitionId partitionId : staticClusterMap.getAllPartitionIds(null)) {
+      String partitionName = partitionId.toPathString();
+      Map<String, String> partitionProperties = new HashMap<>();
+      partitionProperties.put(ClusterMapUtils.PARTITION_STATE,
+          partitionId.getPartitionState() == PartitionState.READ_WRITE ? ClusterMapUtils.READ_WRITE_STR
+              : ClusterMapUtils.READ_ONLY_STR);
+      partitionOverrideInfos.put(partitionName, partitionProperties);
+    }
+    return partitionOverrideInfos;
+  }
+
+  /**
+   * Uploads the seal state of all partitions in the format of map.
+   */
+  private void uploadPartitionOverride(Map<String, Map<String, String>> partitionOverrideInfos) {
+    Properties storeProps = new Properties();
+    storeProps.setProperty("helix.property.store.root.path", "/" + clusterName);
+    HelixPropertyStoreConfig propertyStoreConfig = new HelixPropertyStoreConfig(new VerifiableProperties(storeProps));
+    info("Setting partition override for all datacenters.");
+    for (Map.Entry<String, ClusterMapUtils.DcZkInfo> entry : dataCenterToZkAddress.entrySet()) {
+      HelixPropertyStore<ZNRecord> helixPropertyStore =
+          ClusterMapUtils.createHelixPropertyStore(entry.getValue().getZkConnectStr(), propertyStoreConfig, null);
+      ZNRecord znRecord = new ZNRecord(ClusterMapUtils.ZNODE_NAME);
+      znRecord.setMapFields(partitionOverrideInfos);
```

Comment:
One problem with the map fields is that you can only have one kind of mapping on the znode. Basically mapping on this znode can only be used for partition override.

Reply:
Currently, it is not a serious problem because the znode is purposely created for partition override map. If there is another map later, we could place it in another znode.